mailbox/plugin/deleted-messages-vault-cassandra/src/main/java/org/apache/james/vault/metadata/MetadataDAO.java

```diff
@@ -0,0 +1,146 @@
+ ****************************************************************/
+
+package org.apache.james.vault.metadata;
+
+import static com.datastax.driver.core.querybuilder.QueryBuilder.bindMarker;
+import static com.datastax.driver.core.querybuilder.QueryBuilder.delete;
+import static com.datastax.driver.core.querybuilder.QueryBuilder.eq;
+import static com.datastax.driver.core.querybuilder.QueryBuilder.insertInto;
+import static com.datastax.driver.core.querybuilder.QueryBuilder.select;
+import static org.apache.james.vault.metadata.DeletedMessageMetadataModule.DeletedMessageMetadataTable.BUCKET_NAME;
+import static org.apache.james.vault.metadata.DeletedMessageMetadataModule.DeletedMessageMetadataTable.MESSAGE_ID;
+import static org.apache.james.vault.metadata.DeletedMessageMetadataModule.DeletedMessageMetadataTable.OWNER;
+import static org.apache.james.vault.metadata.DeletedMessageMetadataModule.DeletedMessageMetadataTable.PAYLOAD;
+import static org.apache.james.vault.metadata.DeletedMessageMetadataModule.DeletedMessageMetadataTable.TABLE;
+
+import org.apache.james.backends.cassandra.utils.CassandraAsyncExecutor;
+import org.apache.james.blob.api.BucketName;
+import org.apache.james.core.User;
+import org.apache.james.mailbox.model.MessageId;
+import org.apache.james.vault.dto.DeletedMessageWithStorageInformationConverter;
+import org.apache.james.vault.dto.DeletedMessageWithStorageInformationDTO;
+
+import com.datastax.driver.core.PreparedStatement;
+import com.datastax.driver.core.Session;
+import com.fasterxml.jackson.annotation.JsonInclude;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.datatype.jdk8.Jdk8Module;
+import com.github.fge.lambdas.Throwing;
+
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+
+class MetadataDAO {
+    private final CassandraAsyncExecutor cassandraAsyncExecutor;
+    private final PreparedStatement addStatement;
+    private final PreparedStatement removeStatement;
+    private final PreparedStatement removeAllStatement;
+    private final PreparedStatement readStatement;
+    private final PreparedStatement readMessageIdStatement;
+    private final MessageId.Factory messageIdFactory;
+    private final ObjectMapper objectMapper;
+    private final DeletedMessageWithStorageInformationConverter dtoConverter;
+
+
+    MetadataDAO(Session session, MessageId.Factory messageIdFactory, DeletedMessageWithStorageInformationConverter dtoConverter) {
+        this.cassandraAsyncExecutor = new CassandraAsyncExecutor(session);
+        this.addStatement = prepareAdd(session);
+        this.removeStatement = prepareRemove(session);
+        this.removeAllStatement = prepareRemoveAll(session);
+        this.readStatement = prepareRead(session, PAYLOAD);
+        this.readMessageIdStatement = prepareRead(session, MESSAGE_ID);
+        this.messageIdFactory = messageIdFactory;
+        this.dtoConverter = dtoConverter;
+        this.objectMapper = new ObjectMapper()
```

Comment:
how are you supposed to know what's required to serialize your DTO here? It looks like another responsibility

Reply:
IMO the DAO is responsible of offering a simple API for storing domain objects, leveraging a single Cassandra table, hence the responsibilities felt allright.

Reply:
offering an API is not the same as "implementing everything here". We can very well delegate the serialization to a another component, it would not change the API.

Reply:
We can - but putting the serialization at the DAO scope makes real sense to me.

I prefer that rather than have it "leak" a JSON string, and serializing/deserializing that string in the already complex vault - I struggle to see the value it adds.

Reply:
It's the wrong debate. I just ask for the json (de)serialization to be handled by another class (that my be injected here instead of creating an ObjectMapper).
It's SRP.

Reply:
Let's SRP then. In my eye, this was a small responsibility.